common/expression/src/kernels/take.rs

```diff
@@ -0,0 +1,125 @@
+                Value::Column(c) => after_columns.push(Value::Column(Column::take(c, indices))),
+            }
+        }
+        Ok(Chunk::new(after_columns, indices.len()))
+    }
+}
+
+impl Column {
+    pub fn take<I: Index>(&self, indices: &[I]) -> Self {
+        let length = indices.len();
+        with_number_type!(SRC_TYPE, match self {
+            Column::SRC_TYPE(values) => {
+                Column::SRC_TYPE(Self::take_primitives(values, indices))
+            }
+            Column::Null { .. } | Column::EmptyArray { .. } => self.slice(0..length),
+
+            Column::Boolean(bm) => Self::take_scalars::<BooleanType, _>(
+                bm,
+                MutableBitmap::with_capacity(length),
+                indices
+            ),
+            Column::String(column) => Self::take_scalars::<StringType, _>(
+                column,
+                StringColumnBuilder::with_capacity(length, 0),
+                indices
+            ),
+            Column::Array(column) => {
+                let mut builder = ArrayColumnBuilder::<AnyType>::from_column(column.slice(0..0));
+                builder.reserve(length);
+                Self::take_scalars::<ArrayType<AnyType>, _>(column, builder, indices)
+            }
+            Column::Nullable(c) => {
+                let column = c.column.take(indices);
+                let validity = Self::take_scalars::<BooleanType, _>(
+                    &c.validity,
+                    MutableBitmap::with_capacity(length),
+                    indices,
+                );
+                Column::Nullable(Box::new(NullableColumn {
+                    column,
+                    validity: BooleanType::try_downcast_column(&validity).unwrap(),
+                }))
+            }
+            Column::Tuple { fields, .. } => {
+                let fields = fields.iter().map(|c| c.take(indices)).collect();
+                Column::Tuple {
+                    fields,
+                    len: indices.len(),
+                }
+            }
+        })
+    }
+
+    fn take_scalars<T: ValueType, I: Index>(
+        col: &T::Column,
+        mut builder: T::ColumnBuilder,
+        indices: &[I],
+    ) -> Column {
+        unsafe {
+            for index in indices {
+                T::push_item(
+                    &mut builder,
+                    T::index_column_unchecked(col, index.to_usize()),
+                )
+            }
+        }
+        T::upcast_column(T::build_column(builder))
+    }
+
+    fn take_primitives<T: Copy, I: Index>(col: &Buffer<T>, indices: &[I]) -> Buffer<T> {
+        let mut vs: Vec<T> = Vec::with_capacity(indices.len());
+        let mut dst = vs.as_mut_ptr();
+        for index in indices {
+            unsafe {
+                let e = col[index.to_usize()];
+                dst.write(e);
+                dst = dst.add(1);
+            }
+        }
+        unsafe { vs.set_len(indices.len()) };
```

Comment:
This should be more efficient because `get_unchecked` is used instead of `Index::index`. Besides, `set_len` should have no extra benefit on collection type with `TrustedLen`.

```suggestion
        let mut vs: Vec<T> = indices.map(|index| unsafe { col.get_unchecked(index.to_usize()) }).collect();
```